contribs/gnodev/cmd/gnodev/main.go

```diff
@@ -0,0 +1,360 @@
+`)
+}
+
+func handleDebounce(ctx context.Context, watcher *fsnotify.Watcher, changedPathsCh chan<- []string) error {
+	const timeout = time.Millisecond * 500
+
+	var debounceTimer <-chan time.Time
+	pathList := []string{}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case watchErr := <-watcher.Errors:
+			return fmt.Errorf("watch error: %w", watchErr)
+		case <-debounceTimer:
+			changedPathsCh <- pathList
+			// Reset pathList and debounceTimer
+			pathList = []string{}
+			debounceTimer = nil
+		case evt := <-watcher.Events:
+			if evt.Op != fsnotify.Write {
+				continue
+			}
+
+			pathList = append(pathList, evt.Name)
+			debounceTimer = time.After(timeout)
+		}
+	}
+}
+
+func setupPkgsWatcher(cfg *devCfg, pkgs []gnomod.Pkg) (*fsnotify.Watcher, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("unable to watch files: %w", err)
+	}
+
+	if cfg.noWatch {
+		// noop watcher
+		return watcher, nil
+	}
+
+	for _, pkg := range pkgs {
+		if err := watcher.Add(pkg.Dir); err != nil {
+			return nil, fmt.Errorf("unable to watch %q: %w", pkg.Dir, err)
+		}
+	}
+
+	return watcher, nil
+}
+
+// setupDevNode initializes and returns a new DevNode.
+func setupDevNode(ctx context.Context, logger tmlog.Logger, cfg *devCfg, pkgspath []string, gnoroot string) (*gnodev.Node, error) {
+	if !cfg.minimal {
+		examplesDir := filepath.Join(gnoroot, "examples")
+		pkgspath = append(pkgspath, examplesDir)
+	}
+
+	return gnodev.NewDevNode(ctx, logger, pkgspath)
+}
+
+// setupGnowebServer initializes and starts the Gnoweb server.
+func setupGnowebServer(cfg *devCfg, dnode *gnodev.Node, rt *gnodev.RawTerm) *http.Server {
+	var server http.Server
+
+	webConfig := gnoweb.NewDefaultConfig()
+	webConfig.RemoteAddr = dnode.GetRemoteAddress()
+
+	loggerweb := tmlog.NewTMLogger(rt.NamespacedWriter("GnoWeb"))
+	loggerweb.SetLevel(tmlog.LevelDebug)
+
+	app := gnoweb.MakeApp(loggerweb, webConfig)
+
+	server.ReadHeaderTimeout = 60 * time.Second
+	server.Handler = app.Router
+
+	return &server
+}
+
+func parseArgsPackages(io commands.IO, args []string) (paths []string, err error) {
+	paths = make([]string, len(args))
+	for i, arg := range args {
+		abspath, err := filepath.Abs(arg)
+		if err != nil {
+			return nil, fmt.Errorf("invalid path %q: %w", arg, err)
+		}
+
+		ppath, err := gnomod.FindRootDir(abspath)
+		if err != nil {
+			return nil, fmt.Errorf("unable to find root dir of %q: %w", abspath, err)
+		}
+
+		paths[i] = ppath
+	}
+
+	return paths, nil
+}
+
+func listenForKeyPress(w io.Writer, rt *gnodev.RawTerm) <-chan gnodev.KeyPress {
+	cc := make(chan gnodev.KeyPress, 1)
+	go func() {
+		defer close(cc)
+		key, err := rt.ReadKeyPress()
+		if err != nil {
+			fmt.Fprintf(w, "unable to read keypress: %s\n", err.Error())
+			return
+		}
+
+		cc <- key
+	}()
+
+	return cc
+}
+
+func checkForError(w io.Writer, err error) {
+	if err != nil {
+		fmt.Fprintf(w, "[ERROR] - %s\n", err.Error())
```

Comment:
One annoying thing: if I run `gnodev` and then add a syntax error in a gno file in examples, I have a `| [ERROR] - unable to reset the node: internal error`. Not very helpful.

Reply:
let met check if can catch the underlaying error

Reply:
One suggestion: 

```diff
diff --git a/gno.land/pkg/gnoland/app.go b/gno.land/pkg/gnoland/app.go
index 412930e5..a73b17d4 100644
--- a/gno.land/pkg/gnoland/app.go
+++ b/gno.land/pkg/gnoland/app.go
@@ -158,7 +158,7 @@ func InitChainer(baseApp *sdk.BaseApp, acctKpr auth.AccountKeeperI, bankKpr bank
 
 				// NOTE: comment out to ignore.
 				if !skipFailingGenesisTxs {
-					panic(res.Error)
+					panic(res)
 				}
 			} else {
 				ctx.Logger().Info("SUCCESS:", string(amino.MustMarshalJSON(tx)))
```

`res` contains much more information to find the error than its `Error` field.

Reply:
sounds good to me!

Reply:
Oh actually, I try to use a TmLogger and set log level to error it's not too verbose and errors are now correctly print in the console 👍

Reply:
Nice!